vscode/src/completions/context/context-mixer.ts

```diff
@@ -0,0 +1,169 @@
+import * as vscode from 'vscode'
+
+import { BfgRetriever } from '../../graph/bfg/BfgContextFetcher'
+import { logDebug } from '../../log'
+import { DocumentContext } from '../get-current-doc-context'
+import { ContextRetriever, ContextSnippet } from '../types'
+
+import { JaccardSimilarityRetriever } from './retrievers/jaccard-similarity/jaccard-similarity-retriever'
+import { LspLightGraphCache } from './retrievers/lsp-light/lsp-light-graph-cache'
+
+export interface GetContextOptions {
+    document: vscode.TextDocument
+    position: vscode.Position
+    docContext: DocumentContext
+    abortSignal?: AbortSignal
+    maxChars: number
+}
+
+export type ContextSummary = Readonly<{
+    strategy: string
+    embeddings?: number
+    local?: number
+    graph?: number
+    duration: number
+}>
+
+export interface GetContextResult {
+    context: ContextSnippet[]
+    logSummary: ContextSummary
+}
+
+export class ContextMixer implements vscode.Disposable {
+    private disposables: vscode.Disposable[] = []
+
+    private localRetriever: ContextRetriever | undefined
+    private graphRetriever: ContextRetriever | undefined
+
+    constructor(
+        contextStrategy: 'lsp-light' | 'bfg' | 'jaccard-similarity' | 'none',
+        createBfgRetriever?: () => BfgRetriever
+    ) {
+        if (contextStrategy !== 'none') {
+            this.localRetriever = new JaccardSimilarityRetriever()
+            this.disposables.push(this.localRetriever)
+        }
+
+        if (contextStrategy === 'bfg' && createBfgRetriever) {
+            this.graphRetriever = createBfgRetriever()
+            this.disposables.push(this.graphRetriever)
+        } else if (contextStrategy === 'lsp-light') {
+            this.graphRetriever = LspLightGraphCache.createInstance()
+            this.disposables.push(this.graphRetriever)
+        }
+    }
+
+    // TODO: Generalize the retriever concept more. For now we branch off based on graph context
+    // usage or not to support the existing configuration options
+    public async getContext(options: GetContextOptions): Promise<GetContextResult> {
+        if (this.graphRetriever && this.graphRetriever.isSupportedForLanguageId(options.document.languageId)) {
+            return this.getGraphContext(options)
+        }
+        return this.getLocalContext(options)
+    }
+
+    public async getLocalContext(options: GetContextOptions): Promise<GetContextResult> {
+        const { maxChars } = options
+        const start = performance.now()
+
+        const localMatches =
+            (await this.localRetriever?.retrieve({
+                ...options,
+                hints: {
+                    maxChars: options.maxChars,
+                    maxMs: 150,
+                },
+            })) ?? []
```

Comment:
Redundant because we always return an array?
```suggestion
        const localMatches = await this.localRetriever?.retrieve({
            ...options,
            hints: {
                maxChars: options.maxChars,
                maxMs: 150,
            },
        })
```

Reply:
`localRetriever` can be undefined if the context stragey is `off` (for tests) so that's why i added the `?? []`